test/gradualDeliveryCrowdsale.js

```diff
@@ -0,0 +1,113 @@
+        ],
+    },
+    async ({ contractName, getAccount, fundOwner, getFund, getToken }) => {
+        async function addToWhitelist(...contributors) {
+            if (contractName === "CarryTokenPresale") {
+                const fund = getFund();
+                await fund.addManyToWhitelist(
+                    contributors,
+                    {from: fundOwner}
+                );
+            }
+        }
+
+        it(
+            "should not withdraw tokens immediately after purchase",
+            async function () {
+                const contributor = getAccount();
+                const fund = getFund();
+                const token = getToken();
+                const prevBalance = await token.balanceOf(contributor);
+                await addToWhitelist(contributor);
+                await fund.sendTransaction({
+                    value: web3.toWei(100, "finney"),
+                    from: contributor,
+                });
+                const nextBalance = await token.balanceOf(contributor);
+                assertEq(
+                    prevBalance,
+                    nextBalance,
+                    "Token must not be withdrawn imediately after purchase"
+                );
+            }
+        );
+
+        it("disallows withdrawal by other than the owner", async function () {
+            const nonOwner = getAccount();
+            const fund = getFund();
+            await assertFail(
+                fund.deliverTokensInRatio(1, 2, { from: nonOwner }),
+                "Withdrawal should be failed due to the lack of permission"
+            );
+        });
+
+        it("delivers tokens in the specified ratio", async function () {
+            const contributors = [getAccount(), getAccount(), getAccount()];
+            const fund = getFund();
+            const token = getToken();
+            const initialBalancess =
+                await Promise.all(contributors.map(c => token.balanceOf(c)));
+            await addToWhitelist(...contributors);
+            await fund.sendTransaction({
+                value: web3.toWei(100, "finney"),
+                from: contributors[0],
+            });
+            await fund.sendTransaction({
+                value: web3.toWei(500, "finney"),
+                from: contributors[1],
+            });
+            // contributors[2] contributes 0 ETH
+            const intermediateBalances =
+                await Promise.all(contributors.map(c => token.balanceOf(c)));
+            for (let i = 0; i < initialBalancess.length; i++) {
+                assertEq(
+                    initialBalancess[i], intermediateBalances[i],
+                    "Token must not be withdrawn imediately after purchase"
+                );
+            }
+            await fund.deliverTokensInRatio(1, 2, {from: fundOwner});
+            const finalBalances =
+                await Promise.all(contributors.map(c => token.balanceOf(c)));
```

Comment:
All `initialBalancess`, `intermediateBalances`, and `finalBalances` seems to be zero. Am I understanding something incorrectly?

Reply:
It's about Truffle's half-baked fixture system; it doesn't provide a blank slate for each test case, but shares the whole blockchain state among all test cases.  Their initial balances may be non-empty.